### ISSUE TITLE
numerical inaccuracies with new backend

### ISSUE THREAD

**Issue by apulver** [Theano/Theano#5670]
Using the new backend results in significant numerical discrepancies when computing gradients over certain graphs. In the past I had observed slower training with the new backend, but had not been able to come up with a simple way to reproduce the problem until now. Most of the examples from "DeepLearningTutorials" work fine with either backend. 

To reproduce: run the following code with each different backend and look at the output.  

```
import os
device = 'gpu1'
os.environ['THEANO_FLAGS'] = "floatX=float32, device={}".format(device)


import theano
import theano.tensor as T
import numpy as np
import pickle

k = T.iscalar("k")
A = T.matrix("A")

np.random.seed(123)
u1 = np.linalg.svd(np.random.randn(500,500))[0].astype(T.config.floatX)
u2 = theano.shared(np.linalg.svd(np.random.randn(500,500))[0].astype(T.config.floatX))


result, updates = theano.scan(fn=lambda prior_result, A: T.dot(u2 +theano.map(lambda x:T.nnet.softmax(x)[0], prior_result)[0],T.tanh(T.dot(A,prior_result))),
                              outputs_info=A,
                              non_sequences=A,
                              n_steps=k)


final_result = result[-1]

power = theano.function(inputs=[A,k], outputs=T.grad(T.nnet.softmax(final_result)[0][0],u2), updates=updates)


z = power(u1,200)

f = open('test' + device + '.pickle', 'wb')
pickle.dump(np.array(z), f)
f.close()

print(np.array(z[0][:10]))


```

Also note that the results are close between device=cpu and device=gpu, but device=cuda returns different values. 

Update: Edited to remove code comments. The code is modified from the theano scan tutorial.

**Comment by gaebor** [comment]
I can only confirm. The results of the `cpu` and `gpu` differ in the 3rd or 4th digit. But with the `cuda` backend it differs already in the 1st or second digit. The `config.nvcc.fastmath`  has no effect on this. The precision (`float32` vs `float64`) has very little effect on this.

**Comment by lamblin** [comment]
I can see there is a difference, but I cannot say if it is outside what can be expected due to different summation orders for instance.
The computation graphs look really similar, disabling cuDNN did not make any difference, nor did `alloc_empty_to_zeros`.
However, `optimizer=fast_compile` seemed to give results close to the CPU and old back-end.

**Comment by lamblin** [comment]
`optimizer_excluding=inplace` seems to give the old results back as well.
I'll try to run with DebugMode, but since the differences are really small, I doubt it can detect anything.
If you want to push further, feel free to enable or disable inplace optimizations one by one. It seems it could be any of:
- scan
- add
- tanh
- setsubtensor
- incsubtensor

**Comment by apulver** [comment]
> I can see there is a difference, but I cannot say if it is outside what can be expected due to different summation orders for instance.

In this case wouldn't we expect better results when floatX=float64?

**Comment by apulver** [comment]
Setting optimizer_excluding=local_inplace_setsubtensor gives the correct results.

**Comment by apulver** [comment]
More notes:

The problem still exists when T.tanh is removed.

There is no substantial difference in results when computing the functions output 
```
T.nnet.softmax(final_result)[0][0]
```
instead of its gradient. So the function output is correct, but the gradient is not.

**Comment by apulver** [comment]
Running the scan for 20 iterations also results in discrepancies in the order 10e-6, so it's not just compounding error

**Comment by abergeron** [comment]
What is your cuda driver version?
Are you using the current master of Theano?

In case those are ok, there may be a problem with the code for inplace setsubtensor.

**Comment by apulver** [comment]
> What is your cuda driver version?
Cuda 8.0

> Are you using the current master of Theano?
 Yes

**Comment by lamblin** [comment]
The issue appears somehow linked to the nested scan.
If I replace `theano.map(lambda x:T.nnet.softmax(x)[0], prior_result)[0]` by `T.nnet.softmax(prior_result)`, I manage to get the "good" results, even without disabling inplace optimizations.
I don't know yet if it is something related to scan that makes it unsafe to operate inplace in the IncSubtensor, or if the IncSubtensor itself is doing something wrong.

**Comment by lamblin** [comment]
Also, disabling C code for GpuIncSubtensor and falling back on perform seems to work.

**Comment by apulver** [comment]
Here's another example that produces inaccuracies on the order of 10e-4 with only 20 iterations.


```
import os
device = 'cuda1'
os.environ['THEANO_FLAGS'] = "floatX=float32, device={},lib.cnmem=0.0".format(device)


import theano
import theano.tensor as T
import numpy as np
import pickle

k = T.iscalar("k")
A = T.matrix("A")

np.random.seed(123)
u1 = theano.shared(np.linalg.svd(np.random.randn(100,100))[0].astype(T.config.floatX), borrow=False)
u2 = theano.shared(np.linalg.svd(np.random.randn(100,100))[0].astype(T.config.floatX), borrow=False)


result, updates = theano.scan(fn=lambda prior_result: T.dot(u2 + theano.map(lambda x:T.nnet.softmax(x)[0], prior_result)[0],T.tanh(T.dot(u1,prior_result))),
                              outputs_info=u1,
                              n_steps=k)

final_result = result[-1]

power = theano.function(inputs=[k], outputs=T.grad(T.nnet.softmax(final_result)[0][0],u1),updates=updates)

z = power(20)

f = open('test' + device + '.pickle', 'wb')
pickle.dump(np.array(z), f)
f.close()

print(np.array(z)[0][:10])

```

**Comment by lamblin** [comment]
I think I got it, I'm preparing a fix.

**Comment by apulver** [comment]
Awesome! Glad to hear it.
